### PR TITLE
Fix #433: Allow multiple InterfaceDescriptionBlocks with pcapng.Writer

### LINKED ISSUE #433: Allow multiple InterfaceDescriptionBlocks with pcapng.Writer
Current version only supports one InterfaceDescriptionBlock, but in
general it is allowed to have more then one.

This patch extends the interface so you can provide alternatively
a sequence of InterfaceDescriptionBlock instances (or anything which is
iterable).
[0001-Allow-multiple-InterfaceDescriptionBlocks-with-pcapng-Writer.zip](https://github.com/kbandla/dpkt/files/2653951/0001-Allow-multiple-InterfaceDescriptionBlocks-with-pcapng-Writer.zip)

### LINKED ISSUE #443: dns resource record pack not working in certain cases
I was trying to parse some dns requests/responses, edit some fields, and then write out the new packets when I noticed that for some inputs the newly written packets were messed up. I took out all my changes to the packet and still saw the same issue 

        mdns = dpkt.dns.DNS (udp.data)
        udp.data = mdns
        p_write.writepkt(eth, ts)

tcpdump -vv -n -r rewrite.pcap
....
bad udp cksum 0x6d58 -> 0xbb39!] 11938 q: A? www.youtube.com. 7/13/6 www.youtube.com. CNAME youtube-ui.l.google.com., youtube-ui.l.google.com. A 172.217.6.46, youtube-ui.l.google.com. A 172.217.6.78, youtube-ui.l.google.com. A 172.217.164.110, youtube-ui.l.google.com. A 216.58.195.78, youtube-ui.l.google.com. A 172.217.0.46, youtube-ui.l.google.com. A 172.217.5.110 ns: . NS f.root-servers.net., . NS m.^@^DM-,M-Y^En^@^@^B^@^A^@^@M-&M-^M^@^T^Af^Lroot-servers^Cnet^@^@^@^B^@^A^@^@.M-^M^@^D^AmM-@M-0^@^@^B^@^A^
....

I think the issue occurs when the first RR has type DNS_CNAME and has the url in bytes in the rdata field
 
RR(name='www.youtube.com', type=5, ttl=43919, rlen=22, rdata=b'\nyoutube-ui\x01l\x06google\xc0\x18', cname='youtube-ui.l.google.com')

Then when dpkt is packing it into bytes it calls pack_rdata which will execute this

        def pack_rdata(self, off, label_ptrs):
            # XXX - yeah, this sux
            if self.rdata:
                return self.rdata

But this will not update label_ptrs so when the next RR object references youtube-ui.l.google.com it will pack the url again and create a ptr offset to the newly packed url and store that in label_ptrs.

b'...
\xc0\x0c\x00\x05\x00\x01\x00\x00\xab\x8f\x00\x16\nyoutube-ui\x01l\x06google\xc0\x18\nyoutube-ui\x01l\x06google\xc0\x18\
...'

I see that when I comment out `return self.rdata` it instead executes `pack_name` and correctly stores the label_ptrs and the newly written pcap looks the same as the source 

A workaround is to just set `rr.rdata = None` for all the packets and `pack_rdata` figures it out.

### PR DESCRIPTION
**Pull request by powerswitch**
Current version only supports one InterfaceDescriptionBlock, but in
general it is allowed to have more then one.

This patch extends the interface so you can provide alternatively a
sequence of InterfaceDescriptionBlock instances (or anything which is
iterable).

This commit was created by the patch provided as ZIP file by
René Liebscher as attachment to the issue.

### CODE REVIEW

File: dpkt/pcapng.py [hunk 1]
```diff
@@ -364,26 +364,31 @@ def __init__(self, fileobj, snaplen=1500, linktype=DLT_EN10MB, shb=None, idb=Non
...
-        idb can be an instance of InterfaceDescriptionBlock(LE)
+        idb can be an instance of InterfaceDescriptionBlock(LE) (or sequence of them)
         """
         self.__f = fileobj
 
         if shb:
             self._validate_block('shb', shb, SectionHeaderBlock)
-        if idb:
-            self._validate_block('idb', idb, InterfaceDescriptionBlock)
+            try:
```
Comment (obormot) [inline dpkt/pcapng.py:373]: indentation issue

File: dpkt/pcapng.py [hunk 2]
```diff
@@ -364,26 +364,31 @@ def __init__(self, fileobj, snaplen=1500, linktype=DLT_EN10MB, shb=None, idb=Non
...
 
         shb can be an instance of SectionHeaderBlock(LE)
-        idb can be an instance of InterfaceDescriptionBlock(LE)
+        idb can be an instance of InterfaceDescriptionBlock(LE) (or sequence of them)
         """
         self.__f = fileobj
 
         if shb:
             self._validate_block('shb', shb, SectionHeaderBlock)
-        if idb:
```
Comment (rliebscher) [inline dpkt/pcapng.py:373]: This line was not deleted in the original patch.

Comment (obormot) [reply dpkt/pcapng.py:373]: It should not be deleted. It should go like:
```
if shb:
    self._validate_block('shb', shb, SectionHeaderBlock)

if idb:
    try:
        for idb_ in idb:
            self._validate_block('idb', idb_, InterfaceDescriptionBlock)
    except TypeError: # not iter
        self._validate_block('idb', idb, InterfaceDescriptionBlock)
        idb = [idb]
    ...
```

Comment (powerswitch) [reply dpkt/pcapng.py:373]: Hi, I'm back again. Added the missing line. Sorry for the delay

**Review by obormot — APPROVED** [review]

**Comment by brifordwylie** [comment]
@powerswitch @rliebscher @obormot So pulling this branch locally, the tests run fine. When I merge master the tests are failing due to the changes pulled in with the merge of #502. So two options:
1) Someone else merges master into this branch and makes sure the test all pass (just run 'tox' at the top level)
2) I'll do it..make a new PR.. and merge...

Also I think I'm going to put some branch protections on master, to avoid stuff like this in future :)

**Comment by powerswitch** [comment]
Hi @brifordwylie, sorry for the long absence. I'm only capable to work on the project rarely. But now I'm on holidays, so I pick it up again.

I'll merge the latest master onto the branch and run the tox tests again, will fix them and push the changes. Again, sorry for the delay

**Comment by brifordwylie** [comment]
@powerswitch @rliebscher looks good, thanks a bunch for the contribution... merging now.

**Pull request merged**